hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+        self.asm.push(self.ip, MasmOp::LteImm(imm));
+    }
+
+    /// Pops an element off the stack, and pushes 1 on the stack if that value is an odd number, else 0.
+    pub fn is_odd(self) {
+        self.asm.push(self.ip, MasmOp::IsOdd);
+    }
+
+    /// Pushes the current value of the cycle counter (clock) on the stack
+    pub fn clk(self) {
+        self.asm.push(self.ip, MasmOp::Clk);
+    }
+
+    /// Pushes 1 on the stack if the element on top of the stack is less than 2^32, else 0.
+    pub fn test_u32(self) {
+        self.asm.push(self.ip, MasmOp::U32Test);
+    }
+
+    /// Pushes 1 on the stack if every element of the word on top of the stack is less than 2^32, else 0.
+    pub fn testw_u32(self) {
+        self.asm.push(self.ip, MasmOp::U32Testw);
+    }
+
+    /// Traps if the element on top of the stack is greater than or equal to 2^32
+    pub fn assert_u32(self) {
+        self.asm.push(self.ip, MasmOp::U32Assert);
+    }
+
+    /// Traps if either of the first two elements on top of the stack are greater than or equal to 2^32
+    pub fn assert2_u32(self) {
+        self.asm.push(self.ip, MasmOp::U32Assert2);
+    }
+
+    /// Traps if any element of the first word on the stack are greater than or equal to 2^32
+    pub fn assertw_u32(self) {
+        self.asm.push(self.ip, MasmOp::U32Assertw);
+    }
+
+    /// Casts the element on top of the stack, `a`, to a valid u32 value, by computing `a mod 2^32`
+    pub fn cast_u32(self) {
+        self.asm.push(self.ip, MasmOp::U32Cast);
+    }
+
+    /// Pops an element, `a`, from the stack, and splits it into two elements, `b` and `c`, each of which are a valid u32 value.
+    ///
+    /// The value for `b` is given by `a mod 2^32`, and the value for `c` by `a / 2^32`. They are pushed on the stack in
+    /// that order, i.e. `c` will be on top of the stack afterwards.
+    pub fn split_u32(self) {
+        self.asm.push(self.ip, MasmOp::U32Split);
+    }
+
+    /// Performs unsigned addition of the top two elements on the stack, `b` and `a` respectively, which
+    /// are expected to be valid u32 values.
+    ///
+    /// The specific behavior of the addition depends on the given `overflow` flags:
+    ///
+    /// * `Overflow::Unchecked` - the addition is performed using the `add` op for field elements, which may
+    /// produce a value that is outside of the u32 range, it is the callers responsibility to ensure that the
+    /// resulting value is in range.
+    /// * `Overflow::Checked` - the operation will trap if either operand, or the result, is not a valid u32
+    /// * `Overflow::Wrapping` - computes the result as `(a + b) mod 2^32`, behavior is undefined if either operand
+    /// is not a valid u32
+    /// * `Overflow::Overflowing` - similar to above, the result is computed as `(a + b) mod 2^32`, however a boolean
+    /// is also pushed on the stack after the result, which is 1 if the result of `a + b` overflowed, else 0.
+    ///
```

Comment:
This comment is duplicated several times, and really belongs in the `Overflow` type (where I assume it already exists - I haven't checked).

Reply:
It doesn't, I can move these docs there though.